actors/market/src/balance_table.rs

```diff
@@ -44,12 +46,17 @@ where
     }
 
     /// Adds token amount to previously initialized account.
-    pub fn add(&mut self, key: &Address, value: &TokenAmount) -> Result<(), HamtError> {
+    pub fn add(&mut self, key: &Address, value: &TokenAmount) -> Result<(), ActorError> {
         let prev = self.get(key)?;
         let sum = &prev + value;
         if sum.is_negative() {
-            Err(format!("New balance in table cannot be negative: {}", sum).into())
-        } else if sum.is_zero() && !prev.is_zero() {
+            return Err(actor_error!(
+                illegal_argument,
```

Comment:
The challenge with moving the exit codes a bit lower will be ensuring consistency with the Go code. I traced this one as an example, and on all the paths I could find this results in ErrIllegalState in the Go implementation. Indeed, I'd say ErrIllegalState is your go-to for anything lower than the top-level actor calls, but those top-level call sites might sometimes be able to adapt into a better exit code.

ErrIllegalArgument is mostly only used at the top level when checking parameters directly provided by the user.

On this pass through I'm not going to check them all, but I'd say the default should be `illegal_state` to get maximal matching with current code. And then we need to decide how much we care about exactly matching the exit codes. We'll need to do so for all practical abort paths in order to maintain sync with mainnet.

Reply:
Yeah, I wouldn't push actor errors down the stack like this as it can be difficult to pick a sane default. Really, the correct error here is likely "assertion failed".

Reply:
We may be able to get away with more implicit conversions and custom error types?